### DISCUSSION TITLE
SPDLOG_COMPILED_LIB and binary size

### DISCUSSION THREAD

**Discussion by chengm349** [gabime/spdlog#2233] [Q&A] [answered]
I created libspdlog.a and when I compiled my code, I set -DSPDLOG_COMPILED_LIB and also linked to libspdlog.a. My application is around 100000000 bytes. If I used my own logging lib, the bindary size  is around 80000000.  -g -O2 was turned on. Just want to check whether I used correctly.

**Comment by tt4g** [comment]
It is not possible to guess from the information provided.

Since spdlog is bundled with fmt, the file size may be larger than a simple logging library. Is the size of `libspdlog.a` 20MB?
Also, do you have the link-time optimization (LTO) option enabled when building `libspdlog.a`?

**Comment by chengm349** [comment]
-rw-r--r-- 1 sgfxbuild FX 2096726 Jun 21  2020 libspdlog.a
How to turn on LTO?

**Reply by tt4g** [reply]
> How to turn on LTO?

Add `CMAKE_INTERPROCEDURAL_OPTIMIZATION=ON` to cmake command arguments.

See: https://stackoverflow.com/questions/31355692/how-do-i-enable-link-time-optimization-lto-with-cmake

**Reply by tt4g** [reply]
> -rw-r--r-- 1 sgfxbuild FX 2096726 Jun 21 2020 libspdlog.a

The file size seems to be only 2MB.

Since spdlog and fmt use templates, the binary size will increase as a result of template instantiation.
If your project requires a large number of template instantiation, the binary size may be large.

**Comment by chengm349** [comment]
$ **cmake320 CMAKE_INTERPROCEDURAL_OPTIMIZATION=ON ..**
**-- The CXX compiler identification is GNU 4.8.5**
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Build spdlog: 1.9.2
-- Looking for C++ include pthread.h
-- Looking for C++ include pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - **Failed**
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- Build type: Release
-- Generating example(s)
-- Generating install
-- Configuring done
-- Generating done
-- Build files have been written to: /fs02/home/mcheng/workspace/spdlog/build

cmake320 to make sure the system use version 3.20 and can I ignore Performing Test CMAKE_HAVE_LIBC_PTHREAD - **Failed**

**Reply by tt4g** [reply]
> cmake320 to make sure the system use version 3.20 and can I ignore Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed

spdlog requires thread supports and pthread is found.

> -- Looking for pthread_create in pthread - found
> -- Found Threads: TRUE

**Comment by chengm349** [comment]
$ cmake320 CMAKE_INTERPROCEDURAL_OPTIMIZATION=ON ..
**-- The CXX compiler identification is GNU 10.2.1**
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /opt/rh/devtoolset-10/root/bin/g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Build spdlog: 1.9.2
-- Looking for C++ include pthread.h
-- Looking for C++ include pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
**-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed**
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- Build type: Release
-- Generating example(s)
-- Generating install
-- Configuring done
-- Generating done
-- Build files have been written to: /fs02/home/mcheng/workspace/spdlog/build

$ make -j
[ 70%] Building CXX object CMakeFiles/spdlog.dir/src/spdlog.cpp.o
[ 70%] Building CXX object CMakeFiles/spdlog.dir/src/stdout_sinks.cpp.o
[ 70%] Building CXX object CMakeFiles/spdlog.dir/src/color_sinks.cpp.o
[ 70%] Building CXX object CMakeFiles/spdlog.dir/src/async.cpp.o
[ 70%] Building CXX object CMakeFiles/spdlog.dir/src/file_sinks.cpp.o
[ 70%] Building CXX object CMakeFiles/spdlog.dir/src/cfg.cpp.o
[ 70%] Building CXX object CMakeFiles/spdlog.dir/src/fmt.cpp.o
[ 80%] Linking CXX static library libspdlog.a
[ 80%] Built target spdlog
[ 90%] Building CXX object example/CMakeFiles/example.dir/example.cpp.o
[100%] Linking CXX executable example
[100%] Built target example

ls -ltra
total 2060
drwxr-xr-x 12 mcheng flexusr    4096 Jan  3 15:33 ..
-rw-r--r--  1 mcheng flexusr     300 Jan  3 15:34 spdlog.pc
-rw-r--r--  1 mcheng flexusr    1301 Jan  3 15:34 spdlogConfig.cmake
-rw-r--r--  1 mcheng flexusr    2742 Jan  3 15:34 spdlogConfigVersion.cmake
-rw-r--r--  1 mcheng flexusr    3967 Jan  3 15:34 CPackConfig.cmake
-rw-r--r--  1 mcheng flexusr    4461 Jan  3 15:34 CPackSourceConfig.cmake
-rw-r--r--  1 mcheng flexusr   19271 Jan  3 15:34 CMakeCache.txt
-rw-r--r--  1 mcheng flexusr   13510 Jan  3 15:34 Makefile
-rw-r--r--  1 mcheng flexusr    4570 Jan  3 15:34 cmake_install.cmake
**-rw-r--r--  1 mcheng flexusr 1486832 Jan  3 15:34 libspdlog.a**
drwxr-xr-x  4 mcheng flexusr     303 Jan  3 15:34 .
drwxr-xr-x  3 mcheng flexusr     102 Jan  3 15:34 example
drwxr-xr-x  7 mcheng flexusr    4096 Jan  3 15:34 CMakeFiles

**Comment by chengm349** [comment, marked as answer]
1.4MB or 20MB? How to get 20MB?

**Reply by tt4g** [reply]
I don't know why the size of the application would increase by 20MB.
If the size of the template instantiation is added to the size of the static library, which is 1.4MB, there is no chance of it reaching 20MB.

**Reply by tt4g** [reply]
> cmake320 CMAKE_INTERPROCEDURAL_OPTIMIZATION=ON

Incorrect command.
Use the `-D` option to define CMake variables.

```diff
-cmake320 CMAKE_INTERPROCEDURAL_OPTIMIZATION=ON ..
+cmake320 -DCMAKE_INTERPROCEDURAL_OPTIMIZATION=ON ..
```

Also, since the build type is not specified, the library will be built with debug build.
In debug build, the binary size will be larger because debug symbols will be generated.

If you want to reduce the binary size, specify `-DCMAKE_BUILD_TYPE=Release` to the CMake command for libraries and applications to do a Release build.

```diff
-cmake320 -DCMAKE_INTERPROCEDURAL_OPTIMIZATION=ON ..
+cmake320 -DCMAKE_BUILD_TYPE=Release -DCMAKE_INTERPROCEDURAL_OPTIMIZATION=ON ..
```

**Comment by chengm349** [comment]
**cmake320 -DCMAKE_BUILD_TYPE=Release -DCMAKE_INTERPROCEDURAL_OPTIMIZATION=ON ..**
-- The CXX compiler identification is GNU 10.2.1
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /opt/rh/devtoolset-10/root/bin/g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Build spdlog: 1.9.2
-- Looking for C++ include pthread.h
-- Looking for C++ include pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- Build type: Release
-- Generating example(s)
-- Generating install
-- Configuring done
-- Generating done
-- Build files have been written to: /fs02/home/mcheng/workspace/spdlog/build

make -j
[ 60%] Building CXX object CMakeFiles/spdlog.dir/src/spdlog.cpp.o
[ 70%] Building CXX object CMakeFiles/spdlog.dir/src/color_sinks.cpp.o
[ 70%] Building CXX object CMakeFiles/spdlog.dir/src/stdout_sinks.cpp.o
[ 70%] Building CXX object CMakeFiles/spdlog.dir/src/file_sinks.cpp.o
[ 70%] Building CXX object CMakeFiles/spdlog.dir/src/async.cpp.o
[ 70%] Building CXX object CMakeFiles/spdlog.dir/src/fmt.cpp.o
[ 70%] Building CXX object CMakeFiles/spdlog.dir/src/cfg.cpp.o
[ 80%] Linking CXX static library libspdlog.a
[ 80%] Built target spdlog
[ 90%] Building CXX object example/CMakeFiles/example.dir/example.cpp.o
[100%] Linking CXX executable example
[100%] Built target example

ls -ltra
total 8204
drwxr-xr-x 12 mcheng flexusr    4096 Jan  4 09:24 ..
-rw-r--r--  1 mcheng flexusr     300 Jan  4 09:25 spdlog.pc
-rw-r--r--  1 mcheng flexusr    1301 Jan  4 09:25 spdlogConfig.cmake
-rw-r--r--  1 mcheng flexusr    2742 Jan  4 09:25 spdlogConfigVersion.cmake
-rw-r--r--  1 mcheng flexusr    3967 Jan  4 09:25 CPackConfig.cmake
-rw-r--r--  1 mcheng flexusr    4461 Jan  4 09:25 CPackSourceConfig.cmake
-rw-r--r--  1 mcheng flexusr   19442 Jan  4 09:25 CMakeCache.txt
-rw-r--r--  1 mcheng flexusr   13510 Jan  4 09:25 Makefile
-rw-r--r--  1 mcheng flexusr    4570 Jan  4 09:25 cmake_install.cmake
**-rw-r--r--  1 mcheng flexusr 4643528 Jan  4 09:25 libspdlog.a**
drwxr-xr-x  4 mcheng flexusr     303 Jan  4 09:25 .
drwxr-xr-x  3 mcheng flexusr     102 Jan  4 09:25 example
drwxr-xr-x  7 mcheng flexusr    4096 Jan  4 09:26 CMakeFiles

Why is the latest size bigger:

**-rw-r--r-- 1 mcheng flexusr 1486832 Jan 3 15:34 libspdlog.a**
**-rw-r--r--  1 mcheng flexusr 4643528 Jan  4 09:25 libspdlog.a**

**Reply by tt4g** [reply]
> Why is the latest size bigger:

This is because  link-time optimization (LTO) is enabled by `CMAKE_INTERPROCEDURAL_OPTIMIZATION=ON`.
LTO adds information to the binary to remove symbols that are not used in the library at link time.

This means that the size of the LTO enabled library will be larger, but the binary size of the application will be smaller because unnecessary symbols will be removed when it is finally linked with the application.
